sourced/ml/repo2/coocc.py

```diff
@@ -0,0 +1,131 @@
+import logging
+import os
+from operator import add
+from uuid import uuid4
+from pyspark.sql import column
+from pyspark.sql.functions import size, col
+
+from modelforge.meta import generate_meta
+from modelforge.model import merge_strings, write_model
+from sourced.ml.engine import create_engine, get_tokens
+from sourced.ml.repo2.base import UastExtractor, Transformer, Cacher, UastDeserializer
+from sourced.ml.repo2.token_map import TokenMapTransformer
+from sourced.ml.token_parser import TokenParser
+
+import itertools
+
+from sourced.ml.bblfsh_meta import Roles, Node
+
+
+class SaveASDFCooccModel(Transformer):
+    def __init__(self, output, tokens_list, **kwargs):
+        super().__init__(**kwargs)
+        self.tokens_list = tokens_list
+        self.output = output
+
+    def __call__(self, sparce_matrix):
+        matrix_count = sparce_matrix.count()
+        rows = sparce_matrix.take(matrix_count)
+
+        mat_row, mat_col, mat_weights = zip(*rows)
+
+        write_model(generate_meta("co-occurrences", (1, 0, 0)),
+                    {"tokens": merge_strings(self.tokens_list),
+                     "matrix": {
+                         "shape": (len(self.tokens_list), len(self.tokens_list)),
+                         "format": "coo",
+                         "data": (mat_weights, (mat_row, mat_col))}
+                     },
+                    self.output)
+
+
+class UASTCooccTransformer(Transformer):
+    def __init__(self, token2index, token_parser, prune_size=1, **kwargs):
+        super().__init__(**kwargs)
+        self.token2index = token2index
+        self.token_parser = token_parser
+
+        # TODO(zurk): implement pruning
+        self.prune_size = prune_size
+
+    def _flatten_children(self, root):
+        ids = []
```

Comment:
This is replaced with `uast_ids_to_bag.py`

Reply:
No, it is not the same. It was in ast2vec as separate thing and implement https://github.com/src-d/ml/blob/master/ast2vec/repo2/coocc.md. Code was here 
https://github.com/src-d/ml/blob/master/ast2vec/repo2/cooccbase.py#L50

Reply:
Please study uast_ids_to_bag.py and remove this completely - it provides the needed generalization. If no idea how, I can show next week.